File: src/main/java/com/pyamsoft/powermanager/app/trigger/DeleteTriggerDialog.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.app.trigger;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AlertDialog;
import com.pyamsoft.powermanager.bus.DeleteTriggerBus;
import com.pyamsoft.powermanager.model.event.DeleteTriggerEvent;
import com.pyamsoft.powermanager.model.sql.PowerTriggerEntry;
import timber.log.Timber;

public class DeleteTriggerDialog extends DialogFragment {

  @NonNull private static final String TRIGGER_NAME = "trigger_name";
  @NonNull private static final String TRIGGER_PERCENT = "trigger_percent";
  @SuppressWarnings("WeakerAccess") int percent;
  private String name;

  @CheckResult @NonNull
  public static DeleteTriggerDialog newInstance(@NonNull PowerTriggerEntry trigger) {
    Bundle args = new Bundle();
    DeleteTriggerDialog fragment = new DeleteTriggerDialog();
    args.putString(TRIGGER_NAME, trigger.name());
    args.putInt(TRIGGER_PERCENT, trigger.percent());
    fragment.setArguments(args);
    return fragment;
  }

  @Override public void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    name = getArguments().getString(TRIGGER_NAME, null);
    percent = getArguments().getInt(TRIGGER_PERCENT, -1);

    if (percent == -1) {
      Timber.e("Invalid percent for DeleteTriggerDialog. Dismiss dialog");
      dismiss();
    }
  }

  @NonNull @Override public Dialog onCreateDialog(Bundle savedInstanceState) {
    return new AlertDialog.Builder(getActivity()).setTitle("Delete Trigger")
        .setMessage("Really delete trigger for: " + name + " [" + percent + "%] ?")
        .setNegativeButton("Cancel", (dialogInterface, i) -> {
          dismiss();
        })
        .setPositiveButton("Okay", (dialogInterface, i) -> {
          dismiss();
          DeleteTriggerBus.get().post(DeleteTriggerEvent.create(percent));
        })
        .create();
  }
}

File: powermanager/src/main/java/com/pyamsoft/powermanager/trigger/DeleteTriggerDialog.kt
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.powermanager.trigger

import android.os.Bundle
import android.support.annotation.CheckResult
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.pyamsoft.powermanager.Injector
import com.pyamsoft.powermanager.databinding.DialogTriggerDeleteBinding
import com.pyamsoft.powermanager.trigger.db.PowerTriggerEntry
import com.pyamsoft.powermanager.uicore.WatchedBottomSheet
import timber.log.Timber
import javax.inject.Inject

class DeleteTriggerDialog : WatchedBottomSheet() {

  private var percent: Int = 0
  private lateinit var name: String

  private lateinit var binding: DialogTriggerDeleteBinding
  @field:Inject internal lateinit var presenter: TriggerDeletePresenter

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    name = arguments.getString(TRIGGER_NAME, null)
    percent = arguments.getInt(TRIGGER_PERCENT, -1)

    if (percent < 0) {
      Timber.e("Invalid percent for DeleteTriggerDialog. Dismiss dialog")
      dismiss()
    }

    Injector.with(context) {
      it.inject(this)
    }
  }

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    binding = DialogTriggerDeleteBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    binding.triggerDeleteMsg.text = "This operation cannot be undone."
    binding.triggerDeleteTitle.text = "Really delete trigger for $percent% ?"
  }

  override fun onStart() {
    super.onStart()

    presenter.clickEvent(binding.triggerDeleteCancel, { dismiss() })
    presenter.deleteTrigger(binding.triggerDeleteConfirm, percent, { dismiss() })
  }

  override fun onStop() {
    super.onStop()
    presenter.stop()
  }

  override fun onDestroy() {
    super.onDestroy()
    presenter.destroy()
  }

  companion object {
    private const val TRIGGER_NAME = "trigger_name"
    private const val TRIGGER_PERCENT = "trigger_percent"

    @JvmStatic @CheckResult fun newInstance(trigger: PowerTriggerEntry): DeleteTriggerDialog {
      val args = Bundle()
      val fragment = DeleteTriggerDialog()
      args.putString(TRIGGER_NAME, trigger.name())
      args.putInt(TRIGGER_PERCENT, trigger.percent())
      fragment.arguments = args
      return fragment
    }
  }
}
